### DISCUSSION TITLE
`Exists` Rule to Avoid `N+1` Query Issue in Array Validation

### DISCUSSION THREAD

**Discussion by Jagadish056** [laravel/framework#53420] [Q&A]
I'm trying to validate an array of `id` values in a Laravel form request, but I want the `exists` check to happen only after all other validation rules have passed. This way, I can ensure that database queries for the `exists` validation are executed last, which should help improve efficiency.

Here’s my current `rules()` method:

    public function rules(): array
    {
        return [
            'id' => ['nullable', 'array'],
            'id.*' => ['required', 'integer', 'exists:users,id']
        ];
    }
The problem with this approach is that it creates an `N+1` query issue, as each `id` in the array triggers a separate `exists` query. To optimize this, I want to defer the `exists` check to the end of the validation process, so I can perform a single database query for all `id` values at once.

My idea is to initially validate each `id` as an integer, then perform the `exists` check at the end, like this:

    public function rules(): array
    {
        return [
            'id' => ['nullable', 'array'],
            'id.*' => ['required', 'integer'],
            'id' => ['exists:users,id']
        ];
    }

To handle this, I’m considering using an `after()` hook in the form request to check all `id` values together, ensuring a single database query only if all other rules are pass.

    public function after(): array
    {
        return [
            function (Validator $validator) {
    
                $values = array_unique($validator->validated()['id'] ?? []);
    
                $query = User::query()->whereIn('id', $values);
    
                if ($values && $query->count() !== count($values)) {
                    $validator->errors()->add('id', 'One or more ID are invalid.');
    
                    Flash::validationException($validator)->push();
                }
            }
        ];
    }

**Is there a way in Laravel to delay the `exists` validation rule like this?**

*Or, is there a more efficient approach to avoid the N+1 query issue?*

**Comment by macropay-solutions** [comment]
Have you tried creating a custom rule?

**Reply by Jagadish056** [reply]
The goal is to validate the `id` field twice — first to ensure it's formatted correctly, then again at the end with a rule that checks against the database. In between, other fields are validated as usual. This way, we can perform database checks last while keeping the traditional validation flow intact.

I'm keeping it simple here, avoiding anything more complex — just what's needed.

**Reply by macropay-solutions** [reply]
You could try to create a custom rule . see https://laravel.com/docs/11.x/validation#using-rule-objects
How you create it for "id.*" would be interesting to find out.

**Reply by Jagadish056** [reply]
I’m not sure what you mean about directing me toward a custom rule. My question is strictly about the ordering of validation rules, not about resolving N + 1 queries — as I’ve already taken care of that, as mentioned earlier.

**Reply by henzeb** [reply]
So, you want to group validation rules in non database checks and database rules. If the first group fails, no database checks are done, if the first group succeeds, then any database checks are done ?

**Reply by Jagadish056** [reply]
Yes, exactly! @henzeb, you’ve got the idea.

**Reply by ghost** [reply]
@Jagadish056 bail and stopOnFirstFailure does not work for this scenario? https://laravel.com/docs/11.x/validation#rule-bail

**Reply by Jagadish056** [reply]
@marius-mcp Nope! Why? Try this [53417](https://github.com/laravel/framework/discussions/53417#discussion-7432025) on validation.

**Reply by ghost** [reply]
@Jagadish056 to avoid autocasting strings to ints, you must use integer rule before the query happens.

**Reply by henzeb** [reply]
Yes, but he wants to check if all id's are integers before he checks if all id's are in the database... That is why bail and stop on first failure won't work.

**Comment by ghost** [comment]
@Jagadish056 Just as @macropay-solutions suggested, you can create a custom object rule that will do what you want in the laravel way.

```php
<?php
 
namespace App\Rules;
 
use Closure;
use Illuminate\Contracts\Validation\ValidationRule;
 
class ExistUserId implements ValidationRule
{
    /**
     * Run the validation rule.
     */
    public function validate(string $attribute, mixed $value, Closure $fail): void
    {
        if (null === $value)) {
            return;
        }

        if (!is_array($value) || [] === $value) {
            $fail('The :attribute must be array not empty.');
            return; // I am not sure if fail closure will throw or not.
        }

        if ($value !== \array_filter($value, fn (mixed $value): bool => is_int($value))) {
            $fail('The :attribute array must contains only integers.');
            return; // I am not sure if fail closure will throw or not.
        }

        if (Users::query()->whereIn('id', $values)->count() !== count(array_unique($values))) {
            $fail('The :attribute array must contains only valid user ids.');
        }
    }
}


public function rules(): array
{
    return [
        'id' => [new ExistUserId()],
    ];
}
``` 

https://laravel.com/docs/11.x/validation#using-rule-objects

The above is valid for L >=9

For 8 https://laravel.com/docs/8.x/validation#using-rule-objects you would need 'bail' + as many rules as messages.

**Comment by dazza-dev** [comment]
I faced the same problem, and since there wasn't a solution available, I decided to create a package: [https://github.com/dazza-dev/Laravel-Batch-Validation](https://github.com/dazza-dev/Laravel-Batch-Validation) that solves it. This is the first version, so it may have some bugs, but it would be great if you could try it out and help improve it further. It works well for my use case, but I hope it can be helpful to the Laravel community.

**Comment by SanderMuller** [comment]
I added batched validation to fix the N+1 issue in https://github.com/SanderMuller/laravel-fluent-validation#batched-database-validation which also offers a fully fluent API as well as up to 100x performance boost for wildcard validation (e.g. items.*.name)
